x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportGetUsersAction.java

```diff
@@ -97,4 +131,35 @@ protected void doExecute(Task task, final GetUsersRequest request, final ActionL
             usersStore.getUsers(usersToSearchFor.toArray(new String[usersToSearchFor.size()]), groupListener);
         }
     }
+
+    private void resolveProfileUids(List<User> users, ActionListener<Map<String, String>> listener) {
+        final List<Subject> subjects = users.stream().map(user -> {
+            if (user instanceof AnonymousUser) {
+                return new Subject(user, Authentication.RealmRef.newAnonymousRealmRef(Node.NODE_NAME_SETTING.get(settings)));
+            } else if (ClientReservedRealm.isReserved(user.principal(), settings)) {
+                return new Subject(user, reservedRealm.realmRef());
+            } else {
+                return new Subject(user, nativeRealmRef);
+            }
+        }).toList();
+
+        profileService.searchProfilesForSubjects(subjects, ActionListener.wrap(resultsAndErrors -> {
+            if (resultsAndErrors.errors().isEmpty()) {
+                assert users.size() == resultsAndErrors.results().size();
+                final Map<String, String> profileUidLookup = resultsAndErrors.results()
+                    .stream()
+                    .filter(t -> Objects.nonNull(t.v2()))
+                    .map(t -> new Tuple<>(t.v1().getUser().principal(), t.v2().uid()))
+                    .collect(Collectors.toUnmodifiableMap(Tuple::v1, Tuple::v2));
+                listener.onResponse(profileUidLookup);
+            } else {
+                final ElasticsearchStatusException exception = new ElasticsearchStatusException(
+                    "failed to retrieve profile for users. please retry without fetching profile uid",
+                    RestStatus.BAD_REQUEST
+                );
+                resultsAndErrors.errors().values().forEach(exception::addSuppressed);
+                listener.onFailure(exception);
+            }
```

Comment:
The API call throws error entirely even when profile uid retrieval errors for a single user. We can definitely be more subtle on error handling. But I feel being strict is a safe start.

Reply:
500 makes sense. Updated as suggested.